rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+- `markAsUnpin({ notification_id: string[] })` - mark notifications with given ids as unpinned.
+- `configure()` - updates user-specific notification settings.
+
+#### LocalRepository
+![image](../images/0014/local_repository.png)
+
+##### Notification model
+Notification model is responsible for:
+- Applying policies to an incoming notification. NS must check whether a recipient unsubscribed from a source before processing to delivery.
+- Enhancing an incoming notification with necessary data: 
+    - `expired_at` if not specified. Kibana allows administrators to set how long notifications will be stored. Falls back to 30 - 90 days be default. 
+- Writing new incoming messages to the Notification storage.
+
+Processing a large number of incoming notifications might slow down Kibana and create a disproportionately high load on some Kibana instances.
+The use of a queue allows us to make the handling of incoming notifications asynchronously.
+Async mechanism lets Kibana to cope with spikes in the number of messages created. The queue implementation is based on the Elasticsearch server. 
+
+The logic is run as a background task. The usage of the single background task:
+- reduces the possibility of read- and write-conflicts
+- allows re-using task_manager built-in features: scheduling, retry, monitoring
+
+If Kibana processes notifications in asynchronous batches, we will need a prioritization mechanism to be sure events
+with a higher severity are delivered first without a significant delay. The prioritization mechanism is based on
+priority property value: the notification with the higher value is processed earlier to provide an experience as close to real-time as possible.
+
+##### Notification storage
+
+NS storage keeps different notification statuses for every user. Most Kibana deployments have [from 20 to 90 users](https://elastic.slack.com/archives/C9097ABGC/p1597162974399900),
+although there are some odd instances with >100k users. So we consider two options at the moment (see the [workflow](https://docs.google.com/document/d/18s-BTHog_oPaQKf871gzuhxkah4pK1GYJM9DCLceaLo/edit#)):
+- Store notification state separately from a notification. State for new notifications created for every user when they change the state.
+```typescript
+interface  NotificationState {
+  notification_id: string;
+  recipient_id: string;
+  read_status: boolean;
+  pinned_status: boolean;
+}
+``` 
+This option consumes less storage space but might lead to expensive read operations due to an additional JOIN call to 
+search for notification state objects. We don’t expect Kibana to read more than 10-20 notifications at once, so the overhead might be acceptable.
+- Create a copy for every notification in a user-specific list of notifications. It’s [the recommended way to store data](https://www.elastic.co/guide/en/elasticsearch/reference/current/joining-queries.html),
+but it leads to duplicating notification messages. 
```

Comment:
It would be useful to do some rough math on the scale of data we'd be duplicating before choosing an option here.

Reply:
My calculation is based on https://www.javamex.com/tutorials/memory/ and https://www.javamex.com/tutorials/memory/object_memory_usage.shtml:
The single notification object:
```typescript
interface Notification {
  recipient_id: string[]; // the single uuid 36 char, so use StringBuilder = 144 bytes + 16 bytes for Array = 88 bytes
  priority?: 'normal' | 'high' | 'critical'; // max. 16 char. = 56 bytes 
  created_at: number; // 8 bytes
  expire_at?: number; // 8 bytes
  is_pinned?: boolean; // 8 bytes
  source_type: string; // max. 16 chars = 56 bytes
  group_id?: string // StringBuilder max 100 char. = 144 bytes
  content: { // 16 bytes
    icon?: { // 16 bytes
      euiIconType: string; // StringBuilder = 144 bytes. URL or inlined SVG can take up to several KB
    } 
    severity?: 'normal' | 'high' | 'critical' // max. 16 char. = 56 bytes 
    message: { // 16 bytes
      translations: { // 16 bytes
        [locale: string]: string; // Unicode, max. 80 char. x 2bytes = 160 bytes x N locales
      }
      arguments: { // 16 bytes
       values: object; // assume max. 100 bytes
       defaultMessage: string; // StringBuilder max 100 char. = 144 bytes
       description: string; // StringBuilder max 100 char. = 144 bytes
      }
    }
    title: { // 16 bytes
      url: string; // url max. 200 char x 2 bytes = 400 bytes.
      translations: { // 16 bytes
        [locale: string]: string; // Unicode, max. 30 char. x 2bytes = 60 bytes x N locales
      }
      arguments: { // 16 bytes
       values: object;  // assume max. 100 bytes
       defaultMessage: string; // StringBuilder max 100 char. = 144 bytes
       description: string; // StringBuilder max 100 char. = 144 bytes
      }
    }
    action: {
      url: string; // url max. 200 char x 2 bytes = 400 bytes.
      translations: { // 16 bytes
        [locale: string]: string; // Unicode, max. 30 char. x 2bytes = 60 bytes x N locales
      }
      arguments: { // 16 bytes
       values: object; // assume max. 100 bytes
       defaultMessage: string; // StringBuilder max 100 char. = 144 bytes
       description: string; // StringBuilder max 100 char. = 144 bytes
      }
    }
  }
}
```
with N = 3 supported locales = 3.7KB